### ISSUE TITLE
SS4.1 Tagging

### ISSUE THREAD

**Issue by nglasl** [silverstripe/silverstripe-versioned#230]
This is a follow up to https://github.com/silverstripe/silverstripe-versioned/issues/216 because the tags for this module are incorrect and need updating. My project is running SS4.1.5, and has pulled down 1.3.2 of this module (since that only asks for SS4, and 1.3.2 is the latest stable version for SS4). However, 1.3.2 of this module throws the error outlined in the issue above because the class wasn't introduced until SS4.2. The dependencies were corrected in 1.3.3, but ~~technically 1.3.3 should have replaced 1.3.2~~ this is where it gets tricky because the problem goes back to 1.3.0.

**Comment by nglasl** [comment]
Digging into this a little further, 1.2.4 seems fine https://github.com/silverstripe/silverstripe-versioned/blob/1.2.4/src/GridFieldArchiveAction.php vs https://github.com/silverstripe/silverstripe-versioned/blob/1.3.0/src/GridFieldArchiveAction.php. However, the 1.2 branch has the requirement change to SS4.2 https://github.com/silverstripe/silverstripe-versioned/commits/1.2. This is confusing, and now I don't know what version I *should* be using :)

**Comment by robbieaverill** [comment]
Hey @nglasl, we have a version lock with SilverStripe framework if you remove three minor versions from it - e.g. SS 4.3 = versioned 1.3. Are you noticing this error message in a unit test build? I've seen that before and we usually work around it by specifying which versioned constraint to use. It can pull in incompatible versions if you have a `^1.0` constraint on versioned somewhere in your project, because Composer will install the latest available version even if you're running SilverStripe 4.1 or 4.2 for example

**Comment by nglasl** [comment]
Yeah I'm using https://github.com/symbiote/silverstripe-queuedjobs/blob/4.3.2/composer.json which ends up pulling this down as a dependency. The error goes away if I explicitly define `"silverstripe/versioned": "1.1.5"`, but the reason for creating the issue is:

- others are bound to encounter this
- it's not clear on the cause until you start digging through file or issue history.

Considering this affects so many versions, there's probably not a lot we can do unless the module dependencies are more strict. Even then, others will encounter this on all existing versions of said module.

**Comment by robbieaverill** [comment]
Technically speaking, I think adding the class was a semantic versioning violation, since PHP cannot parse the class without the interface it implements existing. We could probably patch this easily enough with an early return:

```php
if (!interface_exists(GridField_ActionMenuItem::class)) {
    return;
}
class GridFieldArchiveAction...
```

The underlying problem is still that you end up running silverstripe/versioned on a newer version than the one that is tested with the framework/cms modules. It's not easy to retrospectively constrain upper limits on modules. That said, all of our modules should be backwards compatible, and this is a great example of a change that is not backwards compatible.

Would you like to submit a patch?

**Comment by nglasl** [comment]
Yeah agree, we don't really wanting to be running a newer version as it's likely to introduce other issues. For example, I started using 1.2.X of this module and I encountered a middleware issue (another class that was only introduced in SS4.2). Considering this problem goes so far back, and is already tagged numerous times, I don't think there's much we can do without complicating matters further. Perhaps this needs to be something we add to a troubleshooting section?

**Comment by robbieaverill** [comment]
That's a good idea. I still think it's worth fixing the B/C break for future patch releases though.

You're right that if you pull in new features that are enabled by default, it's difficult to ensure nothing breaks there.

I think the only recommendation from the SilverStripe end would be to use recipes and/or to specify module versions you end up installing explicitly in your composer.json constraints. Again, agree that it's tricky when you don't realise you've installed silverstripe/versioned because it's a dependency of something else.

**Comment by robbieaverill** [comment]
cc @silverstripe/core-team for some thought

**Comment by nglasl** [comment]
Just dropping another note here too while I remember, this isn't specific to the versioned module. I've also ended up with a later `admin` version that's throwing similar class not found errors.

**Comment by robbieaverill** [comment]
Yes you’re right it’s specific to any module that isn’t in your project dependencies or a recipe you require.

I think this is how we should approach this going forwards:

* modules provide the functionality and APIs 
* packages like admin that depend on modules like versioned should enable the functionality - they may have bumped dependency constraints in them to ensure dependencies exist when this is the case
* recipes should specify specific versions of modules to install (this already is true)

This means that you can install later versions of modules on older recipes inadvertently while they will lay dormant until you have a version of admin (for example) or versioned-admin (for example) that actually enables the functionality.

The specific example on this issue is actually a pattern that we are trying to move away from- magic class lookups by interface implementations. The “correct” way it should have been done is to enable it via a YAML config in the admin module (or even in versioned, since (unused YAML config is allowed). This example would be backwards compatible if it had been implemented that way.

Does that make sense? If so I will add a note to the house rules documentation to clarify this expectation. While we can’t reimplement the feature “correctly” (well, we could deprecate it all and reimplement it), we can still fix the BC break here because I think generally speaking it will be an isolated incident.

**Comment by robbieaverill** [comment]
Also, that change in versioned should have come with a dependency bump on framework in the versioned module. It’s too late for that now, but also we shouldn’t always do that because it may restrict people from pulling in new features they want to use in their projects. A combination of bumping dependencies and making things opt in via config should be sufficient I think. Thoughts?

**Comment by sminnee** [comment]
So is the issue that versioned 1.3 requires framework 4.3 but 1.3.0, 1.3.1, and 1.3.2 had inappropriately loose constraints? It's been fixed in the future and "be more careful in the future" is a good general principle but it doesn't help fix pre-existing tags.

I think that in general packages should say what they require. If admin depends on a newer version of framework then it should say so. The recipes are a helpful extra layer of protection, but I don't think it's reasonable to say "may be broken without recipes" – if that's the path we want to go down then we should probably declare the mini-packages experiment failed and roll things back together.

For 1.3.0, 1.3.1, and 1.3.2, it would be worth considering fixing the bug in composer.json, deleting the existing tags, and re-tagging, ensuring that it's only a 1-line fix in composer.json. You might think it's "scary" or "not best practise", but if you're websites down on a Friday you don't say "no deploys on Friday, sorry" – you deploy a fix. ;-)

**Comment by robbieaverill** [comment]
I don't think we need to delete and re-tag, we can [patch the dodgy file](https://github.com/silverstripe/silverstripe-versioned/issues/230#issuecomment-490326323) and do patch releases with that in it which would be pulled into everyones projects when they update next. We shouldn't bump composer constraints if they haven't already been done in that line, otherwise the patch wouldn't work.

**Comment by nglasl** [comment]
The problem with patching the file is that we're not sure how compatible the later version of the module is with SS4.1. So coming back to the example I posed earlier, I rolled the version back to 1.2.X but hit another issue that was similar in nature. I'm not opposed to the configs that you mentioned, but it feels like we should just be following semantic versioning, so 1.2.X and 1.3.X require SS4.2. This is already the case on each branch, but **not** on the tags.

The correct solution here *feels* like what @sminnee is suggesting, in that you re-write 1.2.X and 1.3.X history so they depend on SS4.2 (with a one line commit), and then you re-tag the 1.2.X and 1.3.X line. Now composer will always resolve to the correct version (1.1.5 in this case) without bringing new features in. Either way, this is no longer a problem for me now that I have explicitly defined 1.1.5 in my project, so I'm happy for you guys to use your best judgement on how we approach this from here.

**Comment by robbieaverill** [comment]
I'm really opposed to rewriting history. I think these are failures in process on our end which we can fix with new releases now and avoid in future by improving documentation.

**Comment by sminnee** [comment]
Just never write bugs, eh? It's like we're on the blockchain! ;-)

**Comment by sminnee** [comment]
> I don't think we need to delete and re-tag, we can patch the dodgy file and do patch releases with that in it which would be pulled into everyones projects when they update next. 

If that were true, then Nathan wouldn't have raised this ticket. 1.3.3 was released in March AFAIK and it was an attempt to do what you describe. But 1.3.2 is still getting pilled into Nathan's project.

Another way to fix it would be to release SS 4.1.6 that _conflicts_ with 1.3.x, but it's a bit messy.

The underlying issue is that versioned 1.3.0, 1.3.1, and 1.3.2 are toxic and will pollute projects they're pulled into. We need to figure out how to clean that toxic spill.

**Comment by robbieaverill** [comment]
The issue isn't fixed anywhere yet - [1.3.3 doesn't have a fix for this](https://github.com/silverstripe/silverstripe-versioned/blob/1.3.3/src/GridFieldArchiveAction.php).

**Comment by sminnee** [comment]
That's not where it's supposed to be fixed, this is:
https://github.com/silverstripe/silverstripe-versioned/blob/1.3.3/composer.json#L22

That's why Nathan's getting 1.3.2 in his project and not 1.3.3

**Comment by robbieaverill** [comment]
That "fix" has only made it harder to fix correctly. You have to fix the BC break before you increase the minimum dependency. We can revert the composer change, add an early return to fix the BC break, tag that as a patch release, then put the composer bump back in again once it's correctly B/C.

**Comment by sminnee** [comment]
>  but that doesn't fix anyone's problems for older versions

Hence my concern. I'm not sure that "make versioned 1.3 work with admin 4.0" is a reasonable goal. It may be difficult, increases the number of cases to test, unrelated to semver, and adds little value.

**Comment by robbieaverill** [comment]
It should be backwards compatible otherwise we're breaking semver

**Comment by sminnee** [comment]
let's talk face to face because we're talking past each other.

**Comment by robbieaverill** [comment]
😆 I had a coffee, let me summarise my thoughts again:

The main problem is that:
* the change in versioned 1.2 implements an interface that only exists in framework 4.2
* the dependency constraint on framework was ^4.0
* that should have come with a dependency bump on framework to ^4.2, which was missed
* there's been a patch in 1.3.3 to bump the dependency. This doesn't fix the problem, it only stops it from carrying on in future patch versions
  * we need to revert this, add the early return if the interface doesn't exist, tag a new patch release (in versioned 1.2 and 1.3), then bump the dependency back up again
* the way the feature was implemented could also have been done in a configuration based manner without new interfaces (which breaks against older versions of framework)
  * sometimes this may be necessary to do, but I think we should aim not to do this wherever possible so that people can opt in to using new features in modules where they want to rather than us dictating that they cannot

@sminnee suggested rewriting previous tags to bump their dependencies. This has inherent risks in that it will break people's deployments. I'm confident that my proposal above will fix the issue as well. @nglasl raised concerns that other features may also be introduced in this scenario however, which may also break things. I think generally our backward-compatible-changes-only policy is good, and this shouldn't be a problem.

**Comment by robbieaverill** [comment]
I've spoken with @sminnee in person, and we've agreed on these points:

* We should do the patch fix in the versioned 1.3 line (composer revert, interface exists return, tag, bump composer again).
* Upstream dependencies need to be constrained correctly when changes are made.
* We should not be aiming for core modules to work with other minor versions of other core modules. If it happens for free, ideal, but it shouldn't be an objective.

In order to fix this we'll need to either do a full 4.3.x release, or [break lock step](https://github.com/silverstripe/silverstripe-framework/issues/8970) between core modules in order to tag a 1.3.4 of versioned. The linked RFC will determine which option we take.

In the meantime @nglasl would you mind providing your composer.lock file for us to use to ensure our fix works?

**Comment by nglasl** [comment]
I'm happy with that approach. This is what my requirements are on a fresh project (after restricting the versions to fix the problem):

    "require": {
        "php": ">=7.1.0",
        "silverstripe/recipe-core": "1.1.5",
        "silverstripe/versioned": "1.1.5",
        "silverstripe/admin": "1.1.5",
        "symbiote/silverstripe-queuedjobs": "4.3.2"
    },

**Comment by kinglozzer** [comment]
Good discussion here, I’ll leave the wider meta discussion for the “break lock step” issue, but for fixing this specific problem I agree with the [proposed solution above](https://github.com/silverstripe/silverstripe-versioned/issues/230#issuecomment-490743033):

> We should do the patch fix in the versioned 1.3 line (composer revert, interface exists return, tag, bump composer again)

Sounds like the best solution on balance, I’d also have been in the “nervous about overwriting tags” camp.

> Upstream dependencies need to be constrained correctly when changes are made

Definitely - while I can’t think of a specific example right now, I’m sure we’ve been guilty of this before and got away with it. Does _“check whether a dependency ‘bump’ is required whenever you see ‘depends on framework/123’ in a pull request”_ work as a good rule of thumb?

**Comment by chillu** [comment]
I got geared up for a big rant about the combinatorial complexity of minor releases we can realisticall support, but then read on, and the discussion has turned a direction I agree with ;)

Regarding re-tagging patch releases: I think if your core dependencies break in mysterious ways (like this bug report), it's reasonable to expect a `composer update` as a first step towards diagnosing the issue. Assuming that you've constrained your dependencies well (e.g. using `versioned:^1.3`), it'd pull in a new, fixed patch release automatically, right? That won't automatically work if you're going the conservative route and constraining to something strict like `recipe-cms:4.3.2`, but then you're already used to bumping those constraints manually.

> In order to fix this we'll need to either do a full 4.3.x release, or break lock step between core modules in order to tag a 1.3.4 of versioned.

I think let's aim for a full 4.3 release, to fix the immediate problem, while we work through the relatively complex issue of breaking version lockstep.

**Comment by ScopeyNZ** [comment]
> while we work through the relatively complex issue of breaking version lockstep

I would say breaking lockstep is the relatively _less_ complex issue. We can just say "okay let's break lockstep now" and tag only the relevant repos.

**Comment by maxime-rainville** [comment]
Can I throw a somewhat left field suggestion? This issue might have been avoided if we add an entry in our travis build that tries to install dependency with the `--prefer-lowest` composer flag.

Basically, this tell composer to install the lowest possible version of dependencies. That way if we say that a package need `silverstripe/framework ^4.0` and our PR introduces a dependency on an API that got added in 4.2, the build will fail and we'll know we need to increment the minimum version of framework.

**Comment by robbieaverill** [comment]
Interesting idea! Wouldn't we want to test it both ways though, in a perfect world?

**Comment by maxime-rainville** [comment]
Not sure what you mean by "both ways". If you mean to have a matrix entry that tries to install the latest version of everything, that's already implicitly done because composer will try to fetch the highest version of dependencies by default.

**Comment by robbieaverill** [comment]
Yeah - I see, you're talking about a single build that would run the lowest versions, and the rest would work as normal

**Comment by maxime-rainville** [comment]
I can write an RFC if you want.

Otherwise, is there anything else on this specific issue we can do that isn't covered by https://github.com/silverstripe/silverstripe-framework/issues/8970 ?

**Comment by robbieaverill** [comment]
Yeah we have to actually fix this module 😆 then we can close this issue

**Comment by robbieaverill** [comment]
B/C break PR is here: https://github.com/silverstripe/silverstripe-versioned/pull/232

Once merged:
* Tag new 1.3.x release
* Revert #232 
* Tag another new 1.3.x release

That way the latest compatible/installable version of silverstripe/versioned on SilverStripe 4.0 and 4.1 projects will NOT have a backwards compatibility break.

**Comment by robbieaverill** [comment]
Fix has been tagged in 1.3.4, revert PR at https://github.com/silverstripe/silverstripe-versioned/pull/233

**Comment by robbieaverill** [comment]
Ok everything is merged now. I don't think we need to tag another 1.3.x release with the reverted change in it, since 1.3.4 is compatible with old and new releases so it's not worth its own release for #233.

Thank you to everyone for getting involved with this issue! Fixed now.
